### PR TITLE
SRC frame format detect logic fix and topology add for testing capture via SRC

### PR DESCRIPTION
**Pull request by singalsu**
This patch set fixes corrupt audio in SRC capture if the pipeline needs to convert PCM format from capture DAI format. Check is added to fail prepare() if pipelines requires SRC to convert between frame formats. The earlier version did not handle correctly 16 bit capture from 32 bit configured DAI. A test topology is added since earlier there was no topology to test capture-SRC.

### CODE REVIEW

**Comment by singalsu** [comment]
> Do you still see FW panic?

Yes, this was a different issue. There's still the panic if I capture from a DAI with lower rate than DAI rate. Capturing from 16 kHz with 16 - 48 kHz works OK.

File: tools/topology/sof-apl-src-dmic.m4 [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+dnl     pipe id, dai type, dai_index, dai_be,
+dnl     buffer, periods, format,
+dnl     deadline, priority, core, time_domain)
+
+# capture DAI is DMIC 0 using 3 periods
+# Buffers use s32le format, 1000us deadline on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	6, DMIC, 0, NoCodec-6,
+	PIPELINE_SINK_6, 3, s32le,
+	1000, 0, 0)
```
Comment (tlauda) [inline tools/topology/sof-apl-src-dmic.m4]: Please update to use `SCHEDULE_TIME_DOMAIN_TIMER`.
Master branch has been updated in the meantime.

File: tools/topology/sof-apl-src-dmic.m4 [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+# capture DAI is DMIC 1 using 3 periods
+# Buffers use s32le format, 1000us deadline on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	7, DMIC, 1, NoCodec-7,
+	PIPELINE_SINK_7, 3, s32le,
+	1000, 0, 0)
```
Comment (tlauda) [inline tools/topology/sof-apl-src-dmic.m4]: Please update to use `SCHEDULE_TIME_DOMAIN_TIMER`.
Master branch has been updated in the meantime.

File: tools/topology/sof-apl-src-dmic.m4 (hunk 1 above)
Comment (singalsu) [reply tools/topology/sof-apl-src-dmic.m4]: Thanks, done!

File: tools/topology/sof-apl-src-dmic.m4 (hunk 2 above)
Comment (singalsu) [reply tools/topology/sof-apl-src-dmic.m4]: Yep.

**Comment by singalsu** [comment]
SOFCI TEST

File: tools/topology/sof/pipe-src-capture.m4 [hunk 3]
```diff
@@ -2,62 +2,110 @@
...
+W_PCM_CAPTURE(PCM_ID, SRC Capture, 0, 3)
 
 #
 # SRC Configuration
 #
 
-W_VENDORTUPLES(media_src_tokens, sof_src_tokens, LIST(`		', `SOF_TKN_SRC_RATE_IN	"PIPELINE_RATE"'))
+# Create unique names for data to allow several different input rate instances
+define(CAPTURE_SRC_TOKENS, concat(`capture_src_tokens_', PIPELINE_ID))
+define(CAPTURE_SRC_CONF, concat(`capture_src_conf_', PIPELINE_ID))
```
Comment (singalsu) [inline tools/topology/sof/pipe-src-capture.m4]: There's no undefine() for these macros. They are not needed outside this macro so better to do it.

Comment (singalsu) [reply tools/topology/sof/pipe-src-capture.m4]: Done now (undefines add plus minor topology cleanup). I also created separate pipeline building block macros pipe-src-capture.m4 and pipe-src-volume-capture.m4. The latter is more generic with format convert capability. The other ASRC PR has similar pipeline macros.

**Pull request merged**